src/ContaoServices/ServiceFactory.php

```diff
@@ -0,0 +1,163 @@
+     */
+    public function createEnvironmentService()
+    {
+        return $this->container->get('contao.framework')->createInstance(Environment::class);
+    }
+
+    /**
+     * Create the user service for contao user.
+     *
+     * @return BackendUser|FrontendUser
+     *
+     * @throws \RuntimeException Throw an exception if contao mode not defined.
+     */
+    public function createUserService()
+    {
+        $config = $this->container->get('cca.legacy_dic.contao_config');
+        // Work around the fact that \Contao\Database::getInstance() always creates an instance,
+        // even when no driver is configured (Database and Config are being imported into the user class and there-
+        // fore causing an fatal error).
+        if (!$config->get('dbDatabase')) {
+            throw new \RuntimeException('Contao Database is not properly configured.');
+        }
+
+        $matcher = $this->container->get('contao.routing.scope_matcher');
+        $request = $this->container->get('request_stack')->getCurrentRequest();
+
+        // NULL request => CLI mode.
+        if ((null === $request) || $matcher->isBackendRequest($request)) {
+            return $this->container->get('contao.framework')->createInstance(BackendUser::class);
+        }
+
+        if ($matcher->isFrontendRequest($request)) {
+            return $this->container->get('contao.framework')->createInstance(FrontendUser::class);
+        }
+
+        throw new \RuntimeException('Unknown TL_MODE encountered', 1);
+    }
+
+    /**
+     * Create the database connection service for contao database.
+     *
+     * @return Database
+     */
+    public function createDatabaseConnectionService()
+    {
+        // Ensure the user is loaded before the database class.
+        $this->container->get('cca.legacy_dic.contao_user');
+
+        return $this->container->get('contao.framework')->createInstance(Database::class);
+    }
+
+    /**
+     * Create the input service for contao input.
+     *
+     * @return Input
+     */
+    public function createInputService()
+    {
+        return $this->container->get('contao.framework')->createInstance(Input::class);
+    }
+
+    /**
+     * Create the session service for contao session.
+     *
+     * @return Session
+     */
+    public function createSessionService()
+    {
+        return $this->container->get('contao.framework')->createInstance(Session::class);
+    }
+
+    /**
+     * Create the page provider service for provide the current active page model.
+     *
+     * @return PageProvider
+     *
+     * @SuppressWarnings(PHPMD.Superglobals)
+     */
+    public function createPageProviderService()
+    {
+        $pageProvider = new PageProvider();
+
+        if (isset($GLOBALS['TL_HOOKS']['getPageLayout']) && is_array($GLOBALS['TL_HOOKS']['getPageLayout'])) {
+            array_unshift(
+                $GLOBALS['TL_HOOKS']['getPageLayout'],
+                [PageProvider::class, 'setPage']
```

Comment:
We could use the service `cca.legacy_dic.contao_page_provider` instead of the class name.

Reply:
Will you change it?